lib/spack/spack/solver/asp.py

```diff
@@ -2070,6 +2111,9 @@ def setup(self, driver, specs, reuse=None):
         self.add_concrete_versions_from_specs(specs, version_provenance.spec)
         self.add_concrete_versions_from_specs(dev_specs, version_provenance.dev_spec)
```

Comment:
I'm reading through this and I still don't 100% understand why the transformations are necessary.

Is it not sufficient to add another invocation of `add_concrete_versions_from_specs` here, and to pass in a list of the specs from requirements?  I don't see why the other parts of this PR are necessary.

Reply:
Initially, that's what I had, and you can for example see an experimental commit I added which reverts the other stuff: https://github.com/spack/spack/pull/35057/commits/a99e5b273b11dd6f286000f8d27d8d47318cace7. If you check that out and run the tests, then the test `test_requirement_adds_multiple_new_versions` will fail at

```
s2 = Spec("v@{0}".format(commits[1])).concretized()
```

with 

```
    def handle_error(self, msg, *args):
        """Handle an error state derived by the solver."""
        msg = msg.format(*args)
    
        # For variant formatting, we sometimes have to construct specs
        # to format values properly. Find/replace all occurances of
        # Spec(...) with the string representation of the spec mentioned
        specs_to_construct = re.findall(r"Spec\(([^)]*)\)", msg)
        for spec_str in specs_to_construct:
            msg = msg.replace("Spec(%s)" % spec_str, str(spack.spec.Spec(spec_str)))
    
        # TODO: this raises early -- we should handle multiple errors if there are any.
>       raise UnsatisfiableSpecError(msg)
E       spack.solver.asp.UnsatisfiableSpecError: Cannot satisfy the requirements in packages.yaml for the 'v' package. You may want to delete them to proceed with concretization. To check where the requirements are defined run 'spack config blame packages'
```

why is that you might ask? It is because adding versions with `add_concrete_versions_from_specs` does not define the version equivalence between `@hash=number` and the `@hash`. Something needs to do that, or asking for `v@hash` will fail.

Although package requirements are added as imposed specs, they do not get added as version constraints in `define_version_constraints` because

```
(Pdb) x = Spec("v@{0}".format(commits[1]))
(Pdb) y = Spec("v@{0}=2.2".format(commits[1]))
(Pdb) z = Spec("v@2.2")
(Pdb) x.satisfies(y)
False
(Pdb) y.satisfies(x)
False
(Pdb) z.satisfies(y)
False
(Pdb) y.satisfies(z)
True
```

i.e. a hash would never be considered to satisfy a hash=version in the concretizer.

https://github.com/spack/spack/pull/35681 relates to this, although I would need to investigate it to check whether it's considered the case that it would allow `v@hash` to satisfy `v@hash=version`.

This problem would also "go away" if we forbid users from asking for `spec@hash` on the command line (`spec@version` would be OK). However,

* This doesn't seem strictly correct (how can the git commit by itself not satisfy?)
* The error message is quite vague, so anyone responding to someone's bug report would have to be familiar with this idiosyncrasy or risk spending an unpleasant and lengthy span of time to learn this

Reply:
> https://github.com/spack/spack/pull/35681 relates to this, although I would need to investigate it to check whether it's considered the case that it would allow v@hash to satisfy v@hash=version

In that PR I didn't fix the semantic of this case, or at least I didn't yet. To be consistent with the subset semantic for `satisfies` I think it should work the other way around:

1. `v@hash=version` $\subseteq$ `v@hash`
2. `v@hash=version` $\subseteq$ `v@version`
3. `v@hash` $\cap$ `v@version` = `v@hash=version`

where `$\subseteq$` is `satisfies` and `$\cap$` is `intersects`.

Reply:
@alalazo: can you clear up (3) here?  Is that supposed to be `constrain()`?   `intersects()` returns `bool`.

Reply:
Yes, it's supposed to be `constrain` to be crystal clear `intersects` is returning `False` if:

`v@hash` $\cap$ `v@version` = $\emptyset$

and `True` otherwise

Reply:
@scheibelp thanks for explaining that in detail -- the example helps a lot.  Looking at this, I think:
1. We want to allow the user to specify `@commit` (with no version) either on the CLI or in any YAML file (they're going to do that)
2. We want them to be able to specify `@commit=ver` in any of those places, as well.

Basically we shouldn't restrict what they can write -- if it parses as a version, they should be able to write it without prickliness, which is bad UX.

I also think the root cause of the issue is what you showed -- `satisfies()` isn't doing the right thing, and we should fix the issue there.

Looking at what @alalazo wrote, I *think* his semantics are the right ones.  I would tweak this PR so that:

1. `v@hash=version` $\subseteq$ `v@hash`
2. `v@hash` $\subsetneq$ `v@hash=version`
3. `v@hash=version` $\subseteq$ `v@version`
4. `v@version` $\subsetneq$ `v@hash=version`
5. `v@version` $\cap$ `v@hash` = `v@hash=version`
6. `v@v1` $\cap$ `v@hash=v2` = `v@hash=(` `v1` $\cap$ `v2` `)`

etc.

I think we have a minor issue with this one right now:

2. `v@hash` $\subsetneq$ `v@hash=version`

in that we can currently have concrete specs that have a hash but not a version.  So if you do `spack find @commit=version`, and `foo@commit` is installed, we won't find it.  IMO the solution there is to eventually ensure that concretization assigns a version to every commit, and we never get an installed package with just a commit.  But we can change that part later.  Here, we just want to get the support for commit requirements working.

@alalazo I don't think this breaks anything -- it strictly widens the set of things that can be satisfied, right?  I don't think I'm missing something that needs to be preserved with commit `satisfies()`, right?

Reply:
@tgamblin to make sure, are you saying that,

> `v@hash` $\subsetneq$ `v@hash=version`

*should* be the case, and by that do you mean that `v@hash` should not satisfy `v@hash=version`? If so, then `test_requirement_adds_multiple_new_versions` will continue to fail (at least, so long as we depend on `define_version_constraints` to address this), because it would never add a `.version_satisfies` fact to indicate that `v@hash` would be sufficient.

> IMO the solution there is to eventually ensure that concretization assigns a version to every commit,

how does that work? Should we make up a version? e.g. if the package declares:

```
version(2.0, commit=c2)
version(1.0, commit=c1)
```

and you pick some commit x such that c1 < x < c2, should it generate a number like "1.0.revision_number"? That works ok for a single branch but I'm not sure if that would be guaranteed to work for multiple branches.

Reply:
>  If so, then `test_requirement_adds_multiple_new_versions` will continue to fail

Actually, I'm wrong about this (another piece of logic was interfering with this process and I'll be pushing commits shortly to highly that), although I am actually not sure how this works entirely at the moment.

Reply:
I am confused because if it is not the case that `v@hash` satisfies `v@hash=version`, then we have for instance (i.e. as of the current latest commit in this branch) the following `spack solve --show asp` output in `test_requirement_adds_multiple_new_versions`:


```
version_satisfies("v","ef73aeddeee888a7d5e2d9cc646ec933661f3920","ef73aeddeee888a7d5e2d9cc646ec933661f3920").
version_satisfies("v","ef73aeddeee888a7d5e2d9cc646ec933661f3920","ef73aeddeee888a7d5e2d9cc646ec933661f3920=2.3").
version_satisfies("v","ef73aeddeee888a7d5e2d9cc646ec933661f3920=2.3","ef73aeddeee888a7d5e2d9cc646ec933661f3920=2.3").
version_satisfies("v","fca011772147dbd7c7e1014d88c3baf059a28aa6=2.2","fca011772147dbd7c7e1014d88c3baf059a28aa6=2.2").
```

In particular, we do not have 

```
version_satisfies("v","ef73aeddeee888a7d5e2d9cc646ec933661f3920=2.3","ef73aeddeee888a7d5e2d9cc646ec933661f3920").
```

i.e. that choosing the hash satisfies the constraint that the version is "hash=number", in that case I don't know how the concretizer is producing a solution. I thought it might have something to do with the declaration of `version_equivalent` in `asp.py`, but if I comment that out, `test_requirement_adds_multiple_new_versions` still passes.

Reply:
Oh mystery solved: the resulting spec in that test is

`v@11455396ff5e38a39364436654242f417f114414=2.3%gcc@10.2.1 build_system=generic arch=test-debian6-core2`

i.e. the input spec is `v@hash`, and the output is `v@hash=number`, so in that sense the concretizer "overrides" the user's request with the information from the package requirements.